File: backend/app/schemas/capacity.py
"""Capacity planning response schemas — deterministic, no LLM math."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class CapacityKpis(BaseModel):
    active_projects: int
    owners: int
    overallocated_projects: int
    overloaded_owners: int
    total_fte_demand: float
    owner_capacity_fte: float


class OwnerLoad(BaseModel):
    owner: str
    project_count: int
    fte_demand: float
    capacity_fte: float
    utilization_pct: float
    overloaded: bool
    worst_rag: str | None
    areas: list[str]
    project_ids: list[str]


class AreaLoad(BaseModel):
    area: str
    project_count: int
    fte_demand: float
    overallocated_count: int
    owners: list[str]


class CapacityProjectRow(BaseModel):
    portfolio_id: str
    name: str
    area: str | None
    owner: str | None
    rag_status: str | None
    project_status: str | None
    completion_pct: float | None
    fte_demand: float
    overallocation: bool
    planned_start: date | None
    planned_finish: date | None


class CapacityResponse(BaseModel):
    data_version: str
    kpis: CapacityKpis
    by_owner: list[OwnerLoad]
    by_area: list[AreaLoad]
    projects: list[CapacityProjectRow]

File: frontend/src/lib/api/capacity.ts
import { apiGet } from "./client";

export type CapacityKpis = {
  active_projects: number;
  owners: number;
  overallocated_projects: number;
  overloaded_owners: number;
  total_fte_demand: number;
  owner_capacity_fte: number;
};

export type OwnerLoad = {
  owner: string;
  project_count: number;
  fte_demand: number;
  capacity_fte: number;
  utilization_pct: number;
  overloaded: boolean;
  worst_rag: string | null;
  areas: string[];
  project_ids: string[];
};

export type AreaLoad = {
  area: string;
  project_count: number;
  fte_demand: number;
  overallocated_count: number;
  owners: string[];
};

export type CapacityProjectRow = {
  portfolio_id: string;
  name: string;
  area: string | null;
  owner: string | null;
  rag_status: string | null;
  project_status: string | null;
  completion_pct: number | null;
  fte_demand: number;
  overallocation: boolean;
  planned_start: string | null;
  planned_finish: string | null;
};

export type CapacityResponse = {
  data_version: string;
  kpis: CapacityKpis;
  by_owner: OwnerLoad[];
  by_area: AreaLoad[];
  projects: CapacityProjectRow[];
};

export function fetchCapacity(): Promise<CapacityResponse> {
  return apiGet<CapacityResponse>("/api/capacity");
}
